File: code/ex55.py
class Solution:
    def canJump(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        
        if len(nums) == 1:
            return True
        elif nums[0] == 0:
            return False


        for i in range(len(nums)):
            if nums[i] == 0:
                #When i points to the last item in the list which is 0
                if i == len(nums) - 1:
                    break

                recap = 0
                #Recap
                j = i - 1
                while j >= 0:
                    if (nums[j] <= (i - j)):
                        recap += 1

                    j -= 1

                #If there is no valid jump to transcend 0, return false
                if recap == i:
                    return False

        return True


sol = Solution()
print(sol.canJump([2,0,0]))

File: code/ex55.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();    if (n == 1)    return 1;
        for (int i = n - 2; i >= 0; i--) {
            if (nums[i] == 0 && !canOverpass(nums, i))
                return 0;
        }
        return 1;
    }
    
private:
    bool canOverpass(vector<int>& nums, int p) {
        for (int i = p - 1; i >= 0; i--) {
            if (i + nums[i] > p)    return 1;
        }
        return 0;
    }
};